os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,348 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+#include <sys/types.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/timer.h>
+#include <tinyara/irq.h>
+#include "timer_api.h"
+
+#ifdef CONFIG_TIMER
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct amebad_gpt_lowerhalf_s {
+	const struct timer_ops_s *ops;	/* Lowerhalf operations */
+	gtimer_t obj;
+	uint8_t irq_id;
+	bool started;				/* True: Timer is started */
+	tccb_t callback;
+	void *arg;					/* Argument passed to upper half callback */
+	uint32_t gpt_timeout;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+static int amebad_gpt_start(struct timer_lowerhalf_s *lower);
+static int amebad_gpt_stop(struct timer_lowerhalf_s *lower);
+static int amebad_gpt_getstatus(struct timer_lowerhalf_s *lower, struct timer_status_s *status);
+static int amebad_gpt_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout);
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, CODE tccb_t callback, void *arg);
+static int amebad_gpt_ioctl(struct timer_lowerhalf_s *lower, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static const struct timer_ops_s g_timer_ops = {
+	.start = amebad_gpt_start,
+	.stop = amebad_gpt_stop,
+	.getstatus = NULL,
+	.settimeout = amebad_gpt_settimeout,
+	.setcallback = amebad_gpt_setcallback,
+	.ioctl = NULL,
```

Comment:
we need ioctl & get status to get some info & change some modes.

Reply:
Understood. Since in 'timer.c', the getstatus and ioctl functions are marked as optional, only the essential functions are supported in this PR. The getstatus and ioctl will be supported in subsequent update.
Moreover, as so far timer_main.c is the only test case I can use to verify the functions, the ioctl commands used in timer ioctl function (eg. TCIOC_SETDIV) are not included in the test case. Therefore, I am unable to verify them.
Could you please suggest:
1) what ioctl commands need to be supported in this timer ioctl function.
2) the description of those commands and what parameters should I return.
Thanks

Reply:
i'm so sorry, Now I have confirmed this comment.
Could you check imxrt_timer_lowhalf.c ?
If you have any questions, ask me.

Reply:
I think below is necessary.
    1. getstatus to show whether it is started, whether cb is registered, timer resolution (ms or us) and timeleft or current time (or ticks)
    2. set timer resolution (ms or us)
    3. set mode (alarm or freerun)
    4. and something else to control timer

Reply:
Hi @jsdosa, thanks for the reply.

I have the following questions to ask:
1) As we don’t have the option to change the clock divider, resolution, clock source, in this case, what should I add for those ioctl commands?
2) For the free run mode 'CONFIG_EXAMPLES_TIMER_FRT_MEASUREMENT', I would like to know what behaviour do you expect the timer to do?
3) How should I verify those functions after I finished the update. As the timer_main is the only test case I know and I think it does not cover all the tests.

Thanks!

Reply:
1. Does it support both alarm mode and freerun mode? If it is, mode change is necessary.
  Do you mean rtl gpt has fixed resolution? Could you let use know the resolution?
2. That config is usages example of FRT (FreeRun Timer). It just requests getting current time or tick after starting.
  You could find the common behavior by googling of "free run timer" or "free running timer".
3. The timer options are depended on the chipset so that we can't give and cover all of option testing. But the mandatory options can be tested by running timer example which is in apps/examples/timer.

Reply:
> I think below is necessary.
> 1. getstatus to show whether it is started, whether cb is registered, timer resolution (ms or us) and timeleft or current time (or ticks)
> 2. set timer resolution (ms or us)
> 3. set mode (alarm or freerun)
> 4. and something else to control timer

Hi @sunghan-chang ,

Sorry, just now I did not refresh the webpage, so I did not see your reply before I asked the questions.

For your reply,
1. I have done the getstatus function, and it will be committed with the rest of the changes.
2. The resolution of the timer is 32 bits and it shows in us.
3. For the timer mode, currently when I run the timer_main example, the timer is able to perform 10 times of notification before stop. Is it the free run mode behaviour you are excepting? 
4. If so, what behaviour of the alarm mode should it perform? Please advise. 

Many thanks!

Reply:
> 3. For the timer mode, currently when I run the timer_main example, the timer is able to perform 10 times of notification before stop. Is it the free run mode behaviour you are excepting?

It isn't the free-run mode behavior, it's the timer example operation. If you don't give the number of repeat with -n option, it has 10 iterate times by default.
```
In help message
		"  -n [REPEAT]  Number of samples to repeat test.\n" \
		"               DEFAULT 10\n" \

In source code
	if (repeat <= 0) {
		repeat = EXAMPLE_TIMER_NSAMPLES;
	}
```

> 4. If so, what behaviour of the alarm mode should it perform? Please advise.

That's the normal timer. set time and get notification after expired.

Reply:
> 2. The resolution of the timer is 32 bits and it shows in us.

Could you confirm that RTL8721 supports GPT in us only, it does not support ms resolution? It's the same as not support clock source and divider.

Reply:
> Could you confirm that RTL8721 supports GPT in us only, it does not support ms resolution? It's the same as not support clock source and divider.

Rtl8721 uses only the fix clock source of 32kHz, it can provide a tick per 31us. So that the resolution, clock source and divider is fixed. 
Also, as the resolution is 31us, it should be classified as in ms level if that is the meaning of the resolution.

Reply:
Ok. RTL8721 supports the timer in ms.

Reply:
Hi @sunghan-chang @jsdosa ,
As required, the getstatus , ioctl functions as well as the free run feature have been added in the newest update. Please help to review. Many thanks!